flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractNonTimeUnboundedPrecedingOver.java

```diff
@@ -0,0 +1,714 @@
+        if (sortedList == null) {
+            sortedList = new ArrayList<>();
+        }
+        return sortedList;
+    }
+
+    /**
+     * Returns the position of the index where the inputRow must be inserted, updated or deleted.
+     * For insertion, if a suitable position is not found, return -1 to be inserted at the end of
+     * the list. For deletion, if the sortKey is not found, return -1 to indicate the sortKey was
+     * not found in the list.
+     *
+     * @param sortedList
+     * @param inputSortKey
+     * @param isEquals
+     * @return Tuple2(Integer, Boolean) where Integer is the index position in the sortedList and
+     *     Boolean indicates if the element should be inserted or updated
+     */
+    Tuple2<Integer, Boolean> findIndexOfSortKey(
+            List<Tuple2<RowData, List<Long>>> sortedList, RowData inputSortKey, boolean isEquals) {
+        // TODO: Optimize by using Binary Search
+        for (int i = 0; i < sortedList.size(); i++) {
+            RowData curSortKey = sortedList.get(i).f0;
+            if (isEquals && sortKeyEqualiser.equals(curSortKey, inputSortKey)) {
+                return new Tuple2<>(i, true);
+            } else {
+                int compareResult = sortKeyComparator.compare(curSortKey, inputSortKey);
+                if (compareResult == 0) {
+                    // Found inputSortKey
+                    return new Tuple2<>(i, false);
+                } else if (compareResult > 0) {
+                    // Found curSortKey which is greater than inputSortKey
+                    return new Tuple2<>(i, true);
+                }
+            }
+        }
+        // Return not found
+        return new Tuple2<>(-1, true);
+    }
+
+    /**
+     * Helper method to set accumulator and get the value.
+     *
+     * @param accumulator
+     * @return
+     * @throws Exception
+     */
+    RowData setAccumulatorAndGetValue(RowData accumulator) throws Exception {
+        aggFuncs.setAccumulators(accumulator);
+        return aggFuncs.getValue();
+    }
+
+    /**
+     * Helper method to set the accumulator based on the prevIndex.
+     *
+     * @param sortedList
+     * @param prevIndex
+     * @throws Exception
+     */
+    void setAccumulatorOfPrevRow(List<Tuple2<RowData, List<Long>>> sortedList, int prevIndex)
+            throws Exception {
+        if (prevIndex < 0) {
+            aggFuncs.createAccumulators();
+        } else {
+            RowData prevAcc = accMapState.get(sortedList.get(prevIndex).f0);
+            if (prevAcc == null) {
+                aggFuncs.createAccumulators();
+            } else {
+                aggFuncs.setAccumulators(prevAcc);
+            }
+        }
+    }
+
+    /**
+     * Helper method to set the accumulator based on the prevIndex and the ids index j.
+     *
+     * @param sortedList
+     * @param prevIndex
+     * @param idIndex
+     * @throws Exception
+     */
+    void setAccumulatorOfPrevId(
```

Comment:
Should this function be moved to `NonTimeRowsUnboundedPrecedingFunction`?
AFAICT, it is only called from there.